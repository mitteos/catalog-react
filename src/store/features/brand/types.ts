export interface InitialState {
    items: BrandState[]
}

export interface BrandState {
    id: number
    name: string
}
