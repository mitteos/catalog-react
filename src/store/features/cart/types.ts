import {FullProductState} from "store/features/product/types";

export interface InitialState {
    items: {
        product: FullProductState,
        count: number
    }[]
}

export interface AddToCartState {
    product: FullProductState,
    count: number
}

