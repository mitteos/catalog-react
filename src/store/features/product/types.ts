export interface InitialState {
    products: ProductState[]
    selectedType: number | null
    selectedBrand: number[]
    sortValue: "name-decrease" | "name-increase" | "price-decrease" | "price-increase"
    selectedPrice: {min: number; max: number}
    page: number
}
export interface ProductState {
    id: number
    name: string
    img: string
    sizeType: "volume" | "weight"
    sizeValue: number
    barcode: string
    manufacturer: string
    brandId: number
    typeId: number[]
    description: string
    price: number
    vendor: number
}
export interface FullProductState extends ProductState{
    brand: string
    type: string
}

export interface FilterValuesState {
    selectedBrand: number[]
    selectedPrice: {min: number; max: number}
}
