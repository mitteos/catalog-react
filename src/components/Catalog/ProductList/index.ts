export {ProductList} from "components/Catalog/ProductList/ProductList"
