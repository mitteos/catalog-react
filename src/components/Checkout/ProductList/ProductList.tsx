import React from 'react';
import styled from "styled-components";
import {ProductItem} from "components/Checkout";
import { Scrollbars } from 'react-custom-scrollbars';
import {AccentLink} from "components/UI";
import EditIcon from "assets/svg/edit.svg";
import {useAppSelector} from "hooks/redux";

export const ProductList = () => {

    const {items} = useAppSelector(state => state.cart)
    const totalPrice = (items.length && items.map(el => el.product.price * el.count).reduce((prev, next) => prev + next, 0)) || []

    return (
        <Container>
            <Title>Ваш заказ</Title>
            <ScrollContainer
                style={{height: 621}}
                renderTrackVertical={props => <Track {...props} className="track-vertical"/>}
                renderThumbVertical={props => <Thumb {...props} className="thumb-vertical"/>}
                renderView={props => <List {...props} />}
            >
                {items.map(el =>
                        <ProductItem key={el.product.id} product={el.product} count={el.count}/>
                )}
            </ScrollContainer>
            <Footer>
                <Total>
                    <TotalText>ИТОГО</TotalText>
                    <p>{totalPrice} ₸</p>
                </Total>
                <EditBtn href="/cart">
                    <EditText>Редактировать заказ</EditText>
                    <img src={EditIcon} alt=""/>
                </EditBtn>
            </Footer>
        </Container>
    );
};

const Container = styled.div`
  width: 100%;
  padding: 60px 16px 38px 21px;
  background: #FFFFFF;
  box-shadow: 0px 14px 39px -11px rgba(0, 0, 0, 0.06);
  border-radius: 10px;
  @media (max-width: 664px) {
    box-shadow: none;
    padding: 10px 0 0 0;
    background: none;
  }
`
const Title = styled.p`
  font-weight: 700;
  font-size: 30px;
  line-height: 120%;
  color: #111111;
  margin: 0 0 30px;
  @media (max-width: 664px) {
    display: none;
  }
`
const ScrollContainer = styled(Scrollbars)`
  @media (max-width: 664px) {
    display: none;
  }
`
const Thumb = styled.div`
  background-color: #FFC85E;
  width: 6px;
  border-radius: 2px;
  margin-left: -1px;
  display: block;
`
const Track = styled.div`
  position: absolute;
  right: 10px;
  top: 0;
  background-color: rgba(255, 200, 94, 0.4);
  height: 100%;
  width: 4px;
  border-radius: 2px;
  display: block;
  
`
const List = styled.div`
  display: flex;
  flex-direction: column;
  gap: 20px;
  padding: 0 30px 0 0;
`
const Footer = styled.div`
  display: flex;
  flex-direction: column;
  gap: 39px;
  margin: 20px 36px 0 0;
  @media (max-width: 664px) {
    margin: 0;
    flex-direction: row;
    justify-content: space-between;
  }
`
const Total = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
  font-weight: 700;
  font-size: 20px;
  line-height: 120%;
  color: #111111;
  @media (max-width: 664px) {
    font-size: 30px;
  }
`
const TotalText = styled.p`
  @media (max-width: 664px) {
    display: none;
  }
`
const EditBtn = styled(AccentLink)`
  gap: 12px;
  padding: 21px 0;
  width: 237px;
  @media (max-width: 664px) {
    width: 59px;
  }
`
const EditText = styled.p`
  font-weight: 700;
  font-size: 14px;
  line-height: 17px;
  color: #FFFFFF;
  @media (max-width: 664px) {
    display: none;
  }
`
