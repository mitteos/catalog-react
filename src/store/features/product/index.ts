export {productReducer, productActions} from "./productSlice"
export {getAllProducts} from "store/features/product/selectors"
