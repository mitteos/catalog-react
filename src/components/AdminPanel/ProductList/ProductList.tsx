import React from 'react';
import {useAppSelector} from "hooks/redux";
import styled from "styled-components";
import {ProductItem} from "components/AdminPanel";
import {getAllProducts} from "store/features/product";

export const ProductList = () => {

    const products = useAppSelector(getAllProducts)

    return (
        <Container>
            {products.map(product =>
                <ProductItem info={product}/>
            )}
        </Container>
    );
};

const Container = styled.div`
  display: flex;
  flex-direction: column;
  gap: 30px;
`
