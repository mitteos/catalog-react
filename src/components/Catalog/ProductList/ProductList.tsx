import React from 'react';
import styled from "styled-components";
import {ProductItem} from "components/Catalog/ProductItem";
import {useAppSelector} from "hooks/redux";
import {useSearchParams} from "react-router-dom";
import {getAllProducts} from "store/features/product";
import {FullProductState} from "store/features/product/types";
import {Pagination} from "components/Catalog/Pagination";
import {getSortedProducts} from "utils/sortProducts";

interface ProductListProps {
    className?: string
}

export const ProductList: React.FC<ProductListProps> = React.memo(({className}) => {

    const products = useAppSelector<FullProductState[]>(getAllProducts)
    const {selectedType, sortValue, selectedPrice, selectedBrand} = useAppSelector(state => state.product)
    const [searchParams] = useSearchParams()
    const currentPage = searchParams.get("page") || 1
    const queryParamsSearch = searchParams.get("name") || ""
    const productsList = getSortedProducts({
        items: products,
        sort: sortValue,
        page: +currentPage,
        search: queryParamsSearch,
        brand: selectedBrand, type:
        selectedType,
        price: selectedPrice
    })

    return (
        <Container className={className}>
            <List data-testid="product-list">
                {productsList.length
                    ? productsList?.map(product =>
                    <ProductItem key={product.id} info={product} data-testid="product-item"/>)
                    : <Empty data-testid="empty-list">{selectedType && queryParamsSearch ? "В данной категории нет такого товара" : queryParamsSearch ? "Товар с таким названием не найден" : "Товары отсутствуют"}</Empty>
                }
            </List>
            {!!productsList.length &&
                <Pagination />
            }
        </Container>
    );
})

const Container = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 50px;
`
const List = styled.div`
  display: flex;
  flex-wrap: wrap;
  gap: 21px;
  justify-content: center;
`
const Empty = styled.p`
  font-weight: 500;
  font-size: 18px;
  line-height: 130%;
  color: #111111;
`

