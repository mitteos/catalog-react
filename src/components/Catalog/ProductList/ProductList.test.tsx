import {renderWithRouter} from "tests/helpers/renderWithRouter";
import {screen} from "@testing-library/react";
import {FullProductState} from "store/features/product/types";
import {ProductList} from "components/Catalog/ProductList/ProductList";
import * as ReduxHooks from "react-redux";

const products: FullProductState[] = [
    {id: 1, name: "Test 1", price: 100, vendor: 12, type: "test", sizeValue: 12, brand: "test", img: "", sizeType: "volume", typeId: [1], description: "test desc", barcode: "12412", brandId: 1, manufacturer: "test"},
    {id: 2, name: "Test 2", price: 100, vendor: 12, type: "test", sizeValue: 12, brand: "test", img: "", sizeType: "volume", typeId: [1], description: "test desc", barcode: "12412", brandId: 1, manufacturer: "test"},
]

jest.mock("react-redux", () => ({
    ...jest.requireActual("react-redux"),
    useSelector: jest.fn(),
}));

describe("product list", () => {
    test("empty list", () => {
        jest.spyOn(ReduxHooks, "useSelector").mockReturnValue([])
        renderWithRouter(<ProductList />)
        const productList = screen.getByTestId("product-list")
        const emptyMessage = screen.getByTestId("empty-list")
        expect(emptyMessage).not.toBeNull()
        expect(productList).toMatchSnapshot()
    })
})
