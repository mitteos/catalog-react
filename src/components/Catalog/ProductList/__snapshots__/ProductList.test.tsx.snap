// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`product list empty list 1`] = `
<div
  class="sc-hVkBjg fsMZoK"
  data-testid="product-list"
>
  <p
    class="sc-jhSXcr BKjVL"
    data-testid="empty-list"
  >
    Товары отсутствуют
  </p>
</div>
`;
